okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientStream.java

```diff
@@ -139,32 +140,33 @@ void onStreamSentBytes(int numBytes) {
     onSentBytes(numBytes);
   }
 
+  /**
+   * Must be called with holding the transport lock.
+   */
   public void transportHeadersReceived(List<Header> headers, boolean endOfStream) {
-    synchronized (lock) {
-      if (endOfStream) {
-        transportTrailersReceived(Utils.convertTrailers(headers));
-      } else {
-        transportHeadersReceived(Utils.convertHeaders(headers));
-      }
+    if (endOfStream) {
+      transportTrailersReceived(Utils.convertTrailers(headers));
+    } else {
+      transportHeadersReceived(Utils.convertHeaders(headers));
     }
   }
 
-  /**
-   * We synchronized on "lock" for delivering frames and updating window size, because
-   * the {@link #request(int)} call can be called in other thread for delivering frames.
-   */
   public void transportDataReceived(okio.Buffer frame, boolean endOfStream) {
```

Comment:
`transportHeadersReceived()`, `transportDataReceived()`, and `transportReportStatus()` are very similar. I'm quite surprised that two require the lock to be held yet one doesn't. That seems like it is asking for bugs. With my comments below it seems this method should require holding the lock as well.